### PR TITLE
.Net: Instrument the HandleBars planner

### PR DESCRIPTION
**Pull request by TaoChenOSU**
### Motivation and Context

Instrument the Handlebars planner.

### Description

Use the PlannerInstrumention class to wrap the HandleBars planner for instrumentation.
Run using the telemetry sample in the repository:
image
image

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Review by teresaqhoang — APPROVED** [review]

**Comment by teresaqhoang** [comment]
Is HandlebarsPlan.Invoke automatically captured by the function instrumentation?

File: dotnet/src/Planners/Planners.Handlebars/Handlebars/HandlebarsPlanner.cs [hunk 1]
```diff
@@ -46,8 +47,13 @@ public Task<HandlebarsPlan> CreatePlanAsync(Kernel kernel, string goal, Cancella
     {
         Verify.NotNullOrWhiteSpace(goal);
 
-        // TODO (@teresaqhoang): Add instrumentation without depending on planners.core
-        return this.CreatePlanCoreAsync(kernel, goal, cancellationToken);
+        var logger = kernel.GetService<ILoggerFactory>().CreateLogger(this.GetType());
```
Comment (glahaye) [inline dotnet/src/Planners/Planners.Handlebars/Handlebars/HandlebarsPlanner.cs:50]: No need for run-time reflection - Compile-time type determination is possible:

```suggestion
        var logger = kernel.GetService<ILoggerFactory>().CreateLogger(typeof(HandlebarsPlanner));
```

**Review by stephentoub — APPROVED** [review]

File: dotnet/src/InternalUtilities/planning/PlannerInstrumentation.cs [hunk 2]
```diff
@@ -29,9 +29,9 @@ public static class PlannerInstrumentation
 
     /// <summary>Invokes the supplied <paramref name="createPlanAsync"/> delegate, surrounded by logging and metrics.</summary>
     public static async Task<TPlan> CreatePlanAsync<TPlanner, TPlan>(
```
Comment (markwallace-microsoft) [inline dotnet/src/InternalUtilities/planning/PlannerInstrumentation.cs:31]: Does this class show up in the VS Solution Explorer correctly?

File: dotnet/src/InternalUtilities/planning/PlannerInstrumentation.cs [hunk 3]
```diff
@@ -29,9 +29,9 @@ public static class PlannerInstrumentation
 
     /// <summary>Invokes the supplied <paramref name="createPlanAsync"/> delegate, surrounded by logging and metrics.</summary>
```
Comment (markwallace-microsoft) [inline dotnet/src/InternalUtilities/planning/PlannerInstrumentation.cs:30]: See Stephen's PR https://github.com/microsoft/semantic-kernel/pull/3763 where he makes this internal

File: dotnet/src/Planners/Planners.Handlebars/Handlebars/HandlebarsPlanner.cs [hunk 4]
```diff
@@ -46,8 +47,13 @@ public Task<HandlebarsPlan> CreatePlanAsync(Kernel kernel, string goal, Cancella
...
+
+        return PlannerInstrumentation.CreatePlanAsync(
+            static (HandlebarsPlanner planner, Kernel kernel, string goal, CancellationToken cancellationToken)
+                => planner.CreatePlanCoreAsync(kernel, goal, cancellationToken),
+            static (HandlebarsPlan plan) => plan.ToString(),
+            this, kernel, goal, logger, cancellationToken);
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Handlebars/Handlebars/HandlebarsPlanner.cs:56]: Will you also add this to the OpenAI function calling planner

**Review by markwallace-microsoft — APPROVED** [review]

**Review by glahaye — APPROVED** [review]

**Pull request merged**